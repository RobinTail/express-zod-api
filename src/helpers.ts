import {Request} from 'express';
import {z} from 'zod';
import {MiddlewareDefinition} from './middleware';

export type FlatObject = Record<string, any>;
export type ObjectSchema = z.AnyZodObject;

export type Merge<A extends ObjectSchema, B extends ObjectSchema | any> = z.ZodObject<
  // eslint-disable-next-line @typescript-eslint/ban-types
  A['shape'] & (B extends ObjectSchema ? B['shape'] : {}),
  A['_unknownKeys'],
  A['_catchall']
>;

export function combineEndpointAndMiddlewareInputSchemas<IN extends ObjectSchema, mIN>(
  input: IN,
  middlewares: MiddlewareDefinition<any, any, any>[]
): Merge<IN, mIN> {
  if (middlewares.length === 0) {
    return input as any as Merge<IN, mIN>;
  }
  return middlewares
    .map((middleware) => middleware.input)
    .reduce((carry: ObjectSchema, schema) => carry.merge(schema))
    .merge(input) as Merge<IN, mIN>;
}

export function getInitialInput(request: Request): any {
  switch (request.method) {
    case 'POST':
    case 'PUT':
    case 'PATCH':
      return request.body;
    case 'GET':
      return request.query;
    case 'DELETE': // _may_ have body
    default:
      return {...request.query, ...request.body};
  }
}
