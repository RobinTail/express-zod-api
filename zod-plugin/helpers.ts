import type { z } from "zod";
import { unpack } from "./packer";

export const getBrand = (subject: z.core.$ZodType) => {
  const { brand } = unpack(subject) || {};
  if (
    typeof brand === "symbol" ||
    typeof brand === "string" ||
    typeof brand === "number"
  )
    return brand;
  return undefined;
};
