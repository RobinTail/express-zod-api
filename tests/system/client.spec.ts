import { ChildProcessWithoutNullStreams, spawn } from "node:child_process";
import { expectType } from "tsd";
import { config } from "../../example/config";
import { waitFor } from "../helpers";
import {
  ExpressZodAPIClient,
  Implementation,
  jsonEndpoints,
} from "../../example/example.client";

describe("Example", () => {
  let example: ChildProcessWithoutNullStreams;
  let out = "";
  const listener = (chunk: Buffer) => {
    out += chunk.toString();
  };

  beforeAll(() => {
    example = spawn("ts-node", ["example/index.ts"]);
    example.stdout.on("data", listener);
  });

  afterAll(async () => {
    example.stdout.removeListener("data", listener);
    example.kill();
    await waitFor(() => example.killed);
  });

  afterEach(() => {
    out = "";
  });

  const createDefaultImplementation =
    (host: string): Implementation =>
    async (method, path, params) => {
      const searchParams =
        method === "get" ? `?${new URLSearchParams(params)}` : "";
      const response = await fetch(`${host}${path}${searchParams}`, {
        method: method.toUpperCase(),
        headers:
          method === "get" ? undefined : { "Content-Type": "application/json" },
        body: method === "get" ? undefined : JSON.stringify(params),
      });
      if (`${method} ${path}` in jsonEndpoints) {
        return response.json();
      }
      return response.text();
    };

  const client = new ExpressZodAPIClient(
    createDefaultImplementation(`http://localhost:${config.server.listen}`),
  );

  test("Should listen", async () => {
    await waitFor(() => /Listening 8090/.test(out));
    expect(true).toBeTruthy();
  });

  test("Should perform the request with a positive response", async () => {
    const response = await client.provide("get", "/v1/user/retrieve", {
      id: "10",
    });
    expect(response).toMatchSnapshot();
    expectType<
      | { status: "success"; data: { id: number; name: string } }
      | { status: "error"; error: { message: string } }
    >(response);
  });
});
