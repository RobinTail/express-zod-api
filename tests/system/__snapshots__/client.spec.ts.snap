// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example Should perform the request with a positive response 1`] = `
{
  "data": {
    "features": [
      {
        "features": [
          {
            "features": [],
            "title": "Above 180cm",
          },
        ],
        "title": "Tall",
      },
      {
        "features": [],
        "title": "Young",
      },
      {
        "features": [
          {
            "features": [
              {
                "features": [],
                "title": "About Typescript",
              },
            ],
            "title": "Tells funny jokes",
          },
        ],
        "title": "Cute",
      },
    ],
    "id": 10,
    "name": "John Doe",
  },
  "status": "success",
}
`;
